src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/HandlerMethodDetails.cs

```diff
@@ -0,0 +1,185 @@
+            const string HandleRequestSuffix = "." + HandleRequestName;
+            const string HandleNotificationName = nameof(INotificationHandler<object, object>.HandleNotificationAsync);
+            const string HandleNotificationSuffix = "." + HandleNotificationName;
+
+            foreach (var methodInfo in handlerType.GetRuntimeMethods())
+            {
+                if (MethodInfoMatches(methodInfo))
+                    return methodInfo.GetCustomAttribute<LanguageServerEndpointAttribute>();
+            }
+
+            throw new InvalidOperationException("Somehow we are missing the method for our registered handler");
+
+            bool MethodInfoMatches(MethodInfo methodInfo)
+            {
+                var methodName = methodInfo.Name;
+
+                switch (requestType != null, responseType != null)
+                {
+                    case (true, true):
+                        return NameMatches(methodInfo, HandleRequestName, HandleRequestSuffix)
+                            && TypesMatch(methodInfo, [requestType, contextType, typeof(CancellationToken)]);
+
+                    case (false, true):
+                        return NameMatches(methodInfo, HandleRequestName, HandleNotificationSuffix)
+                            && TypesMatch(methodInfo, [contextType, typeof(CancellationToken)]);
+
+                    case (true, false):
+                        return NameMatches(methodInfo, HandleNotificationName, HandleNotificationSuffix)
+                            && TypesMatch(methodInfo, [requestType, contextType, typeof(CancellationToken)]);
+
+                    case (false, false):
+                        return NameMatches(methodInfo, HandleNotificationName, HandleNotificationSuffix)
+                            && TypesMatch(methodInfo, [contextType, typeof(CancellationToken)]);
+                }
+            }
+
+            static bool NameMatches(MethodInfo methodInfo, string name, string suffix)
+            {
+                var methodName = methodInfo.Name;
+
+                return methodName == name || methodName.EndsWith(suffix);
+            }
+
+            static bool TypesMatch(MethodInfo methodInfo, Type?[] types)
+            {
+                var parameters = methodInfo.GetParameters();
+                if (parameters.Length != types.Length)
+                    return false;
+
+                for (int i = 0, n = parameters.Length; i < n; i++)
+                {
+                    if (!Equals(types[i], parameters[i].ParameterType))
+                        return false;
+                }
+
+                return true;
+            }
+        }
+
+        static LanguageServerEndpointAttribute? GetMethodAttributeFromClassOrInterface(Type type)
+        {
+            var attribute = Attribute.GetCustomAttribute(type, typeof(LanguageServerEndpointAttribute)) as LanguageServerEndpointAttribute;
```

Comment:
is it possible (or a good idea) to make this throw if there is a a class level attribute, but implements multiple IMethodHandler interfaces?

Reply:
I'm happy to add some validation, but I'm not going to claim to understand exactly how these attributes are expected to work. :smile: This code was simply moved from here:

https://github.com/dotnet/roslyn/blob/5f778bde9a17c34480dc83f3a17bf4cd9277d630/src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/HandlerProvider.cs#L188-L206

Reply:
if this was just moved then fine with it as-is